src/NUnitEngine/nunit.engine/Services/TestSelectionParser.cs

```diff
@@ -0,0 +1,275 @@
+        private readonly Token OR_OP2 = new Token(TokenKind.Symbol, "||");
+        private readonly Token NOT_OP = new Token(TokenKind.Symbol, "!");
+
+        public string Parse(string input)
+        {
+            if (input == null || input == "")
+                return "";
+
+            _tokenizer = new Tokenizer(input);
+
+            return ParseFilterExpression();
+        }
+
+        /// <summary>
+        /// Parse a single term or an or expression, returning the xml
+        /// </summary>
+        /// <returns></returns>
+        public string ParseFilterExpression()
+        {
+            var terms = new List<string>();
+            terms.Add(ParseFilterTerm());
+
+            while (true)
+            {
+                Token op = Peek();
+                if (op != OR_OP1 && op != OR_OP2)
+                    break;
+
+                NextToken();
+                terms.Add(ParseFilterTerm());
+            }
+
+            if (terms.Count == 1)
+                return terms[0];
+
+            var sb = new StringBuilder("<or>");
+
+            foreach (string term in terms)
+                sb.Append(term);
+
+            sb.Append("</or>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single element or an and expression and return the xml
+        /// </summary>
+        public string ParseFilterTerm()
+        {
+            var elements = new List<string>();
+            elements.Add(ParseFilterElement());
+
+            while (true)
+            {
+                Token op = Peek();
+                if (op != AND_OP1 && op != AND_OP2)
+                    break;
+
+                NextToken();
+                elements.Add(ParseFilterElement());
+            }
+
+            if (elements.Count == 1)
+                return elements[0];
+
+            var sb = new StringBuilder("<and>");
+
+            foreach (string element in elements)
+                sb.Append(element);
+
+            sb.Append("</and>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single filter element such as a category expression
+        /// and return the xml representation of the filter.
+        /// </summary>
+        public string ParseFilterElement()
+        {
+            Token nextOp = Peek();
+            if (nextOp == LPAREN || nextOp == NOT_OP)
+                return ParseExpressionInParentheses();
+
+            string lhs = GetWord();
+
+            switch (lhs)
+            {
+                case "id":
+                    string op = GetOperator("=", "==");
+                    string values = GetIdList();
+                    return string.Format("<id>{0}</id>", values);
+
+                case "cat":
+                case "category":
+                case "method":
+                case "class":
+                    op = GetOperator("=", "==", "!=", "=~", "!~");
```

Comment:
I have simplified the syntax and the code for my next commit. I droped category and fullname and I'm working with Tokens rather than strings, with defined constants for the operator tokens. Thanks for your comments - it looks much better now.